### PR TITLE
Simplify and improve signature verification for workloads on Windows

### PR DESCRIPTION
**Pull request by joeloff**
# Description

Improve and simplify Authenticode verification on Windows for workloads (MSI based installs).

* Add revocation checks when calling `WinVerifyTrust`. Revocation checks default to online, except if the `AllowOnlineRevocationChecks` policy key is set to 0. The policy setting can be used to perform revocation checks when network access is restricted.
* Validate that the certificate chain terminates in a trusted Microsoft root (certificate pinning). Previously this was a weak check relying on the certificate subject. The check now relies on verifying the trusted root chain policy. This isn't exposed by `X509` implementations as it's specific to Windows only.

This change also introduces the use of Microsoft.Windows.CsWin32 to generate all the native structs and API calls. As a result, we can safely remove a chunk of code and rely on the generated source code instead.

This change is a backport from [main](#36787)

# Testing

Unit tests and test assets have been updated. The tests include assets that contain
- tampered files (files with valid signatures that were edited after being signed)
- signed files where the certificate chain terminates in a non-Microsoft root
- signed files containing SHA1 - these are valid, but the root is no longer trusted
- valid SHA2 signatures terminating in a trusted Microsoft root.

### CODE REVIEW

**Comment by joeloff** [comment]
Approved over email, updating the label

**Review by nagilson — APPROVED** [review]
LGTM, I'm glad we got some more subject matter experts to review over at https://github.com/dotnet/sdk/pull/36787/files. These changes are all the same and appropriate for 8.0.

I'm guessing all of our shipped 8.0 material has certificate chains that terminate in a trusted Microsoft root, so this shouldn't break anyone?

**Comment by richlander** [comment]
This change should be tested in containers before merging.

**Comment by joeloff** [comment]
> This change should be tested in containers before merging.

What would make containers different? Do they even use MSI based installations?

File: src/Cli/dotnet/NativeMethods.txt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+// Structs, constants, enums
+WINTRUST_DATA
+WINTRUST_ACTION*
+WTD*
+CERT_STORE*
+CERT_CHAIN_POLICY*
+MICROSOFT_ROOT_CERT*
+TRUST_E*
+CERT_E*
```
Comment (MiYanni) [inline src/Cli/dotnet/NativeMethods.txt:15]: What process uses this file?

File: src/Cli/dotnet/commands/dotnet-workload/SignCheck.cs [hunk 2]
```diff
@@ -2,34 +2,57 @@
...
+        /// Determines whether dotnet.dll is signed.
         /// </summary>
-        /// <returns><see langword="true"/> if dotnet is signed; <see langword="false"/> otherwise.</returns>
-        public static bool IsDotNetSigned() => IsSigned(s_dotnet);
+        /// <returns><see langword="true"/> if dotnet is signed; otherwise, <see langword="false"/>.</returns>
+        public static bool IsDotNetSigned()
+        {
+            if (OperatingSystem.IsWindows())
+            {
+#if !DOT_NET_BUILD_FROM_SOURCE
```
Comment (MiYanni) [inline src/Cli/dotnet/commands/dotnet-workload/SignCheck.cs:29]: Does this mean we don't check the authenticode signature for source builds?

**Comment by richlander** [comment]
You can install MSIs in Windows Server Core containers. We should just validate that we're not breaking anything here.

File: src/Cli/dotnet/NativeMethods.txt (hunk 1 above)
Comment (joeloff) [reply src/Cli/dotnet/NativeMethods.txt:15]: CS/Win32 code generators.

File: src/Cli/dotnet/commands/dotnet-workload/SignCheck.cs (hunk 2 above)
Comment (joeloff) [reply src/Cli/dotnet/commands/dotnet-workload/SignCheck.cs:29]: yup, SB gets unhappy - we don't ship SB copies.

**Comment by joeloff** [comment]
> You can install MSIs in Windows Server Core containers. We should just validate that we're not breaking anything here.

I don't see any issues, sign verification worked fine under a container on my machine using mcr.microsoft.com/windows/servercore:ltsc2022

**Comment by richlander** [comment]
Thanks for checking.

**Pull request merged**